src/blog/konfig-manager/intro-spring-boot.md

```diff
@@ -0,0 +1,334 @@
+---
+Author: Saul Nachman & Moshe Immerman
+Last updated: 23/07/2021
+---
+
+# Konfig Manager
+
+<!-- markdownlint-disable MD036 -->
+**Managing a Spring Boot configuration**
+<!-- markdownlint-enable MD036 -->
+
+The Konfig Manager is for platform engineers needing to manage overlapping configurations across applications and environments.
+
+## Design principles
+
+- **100% YAML** – `Konfigs` are valid YAML and IDE validation and autocomplete of k8s resources works as normal.
+- **Simple** – Easy to use and quick to get started.
+- **Reconciliation based** – Changes are applied quickly and resiliently (unlike webhooks) at runtime.
+
+## Installation
+
+### Prerequisites
+
+This guide assumes you have either a [kind cluster](https://kind.sigs.k8s.io/docs/user/quick-start/) or [minikube cluster](https://minikube.sigs.k8s.io/docs/start/) running, or have some other way of interacting with a cluster via [kubectl](https://kubernetes.io/docs/tasks/tools/).
+
+To test the deployment in browser, set up ingress to your cluster – see, for example, the [minikube](https://kubernetes.io/docs/tasks/access-application-cluster/ingress-minikube/#create-an-ingress-resource) or [kind](https://kind.sigs.k8s.io/docs/user/ingress) docs.
+
+### Install
+
+```bash
+export VERSION=0.10.0
+# For the latest release version: https://github.com/flanksource/konfig-manager/releases
+
+# Apply the operator
+kubectl apply -f https://github.com/flanksource/konfig-manager/releases/download/v${VERSION}/operator.yml
+```
+
+### Following the logs
+
+To follow the konfig-manager logs, open a new terminal and, changing what needs to be changed, run:
+
+```bash
+kubectl logs -f --since 10m -n konfig-manager konfig-manager-d54f64588-hjq5z konfig-manager
+```
+
+These logs are where reconciliation successes and errors show up – and the best place to look when debugging.
+
+## Use case: Externalising a Spring Boot configuration
+
+*As a platform engineer, I need to manage overlapping configurations across multiple applications and environments, so that applications are easy to promote.*
+
+The Java Spring Boot project [recommends developers externalise their configuration](https://docs.spring.io/spring-boot/docs/1.0.1.RELEASE/reference/html/boot-features-external-config.html) and one method it provides is the `application.properties` file.
+
+The Konfig Manager makes it easy to convert a set of `ConfigMaps` and `Secrets` to a single output  – the `application.properties` file – providing a single configuration to the container file system.
+
+The `Konfig` type allows developers to specify a `hierarchy` (defining merge inclusion and precedence) and an `output` field (defining output structure).
+
+### Step 1: Apply a konfig
```

Comment:
I think we should create the Konfig object after we have all the objects CM/Secrets that it references applied. Since before this the operator is already running and as soon as the Konfig object is applied the recocile will trigger and can lead to a fail reconcile loop. So I think it makes sense to replace step1 and step2

Reply:
The downside to that is that the Konfig concept isn't introduced up front, and I think it's the first thing a person needs to know to start making sense of what's going on. 

I don't see a failed reconcile loop as a problem except that it takes a bit of time to get a result then. 

Do you think splitting the different and instructing a rollout restart after applying konfig is a good compromise?

Reply:
I don't think so.... I think we should be creating those resources before hand maybe with a comment...`before creating Konfig lets create some objects which will be feeded to the Konfig` or something similar

Reply:
Could you articulate why you think the downsides outweigh the benefits?

Reply:
The downside is the same as I stated earlier. For me, it makes sense to have all the prereq objs created earlier which would be used by the Konfig.... also this way we'll be sure there will be no reconcile error. Also since we are introducing the konfig-manager we would want the users to always create the required resources before hand and not after they have created the konfig object

Reply:
I get where you're coming from, but I don't think the downside outweigh the benefits. I don't see a problem with a reconcile error as part of the point of using the reconciliation cycle is that objects can be added at any time. The downside for a tutorial, as I see it, is that it sucks a bit to wait, but that can be helped. I've added a rollout restart instruction to help notify the reader. 

As @moshloop has pointed out to me a few times, we need to catch the eye as early as possible in the post. I believe the ordering here does a better job of getting the core mechanism in front of the reader ASAP.

Reply:
In your case since the objects are applied as soon as possible when someone is trying out the blog they might not get the expected result because as soon as the Konfig object is created the reconcile loop is triggered... and if at that moment the referenced object is not there it will just skip them.... Also, suppose you create those objects just after a second as per the guide but since the reconcile loop is also passed that state, the use would have to wait until the next reconcile cycle which can be as long as 10 minutes to get the expected result

Reply:
I've added an instruction to restart the `konfig-manager` so we avoid that issue.

Reply:
Amended as requested.